lib/ipam/ipam.go

```diff
@@ -478,30 +516,117 @@ func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, a
 						break
 					}
 					logCtx.WithError(err).Errorf("Error getting block for affinity")
-					return ips, err
+					return nil, false, err
 				}
 
-				// Claim successful.  Assign addresses from the new block.
-				logCtx.Infof("Claimed new block %v - assigning %d addresses", b, rem)
-				numBlocksOwned++
-				newIPs, err := c.assignFromExistingBlock(ctx, b, rem, handleID, attrs, host, config.StrictAffinity)
-				if err != nil {
-					if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
-						log.WithError(err).Debug("CAS Error assigning from new block - retry")
-						continue
-					}
-					logCtx.WithError(err).Warningf("Failed to assign IPs in newly allocated block")
-					break
+				// Claim successful.
+				block := allocationBlock{b.Value.(*model.AllocationBlock)}
+				if block.NumFreeAddresses() >= minFreeIps {
+					logCtx.Infof("Block '%s' has %d free ips which is more than %d ips required.", b.Key.(model.BlockKey).CIDR, block.NumFreeAddresses(), minFreeIps)
+					return b, true, nil
+				} else {
+					errString := fmt.Sprintf("Block '%s' has %d free ips which is less than %d ips required.", b.Key.(model.BlockKey).CIDR, block.NumFreeAddresses(), minFreeIps)
+					logCtx.Errorf(errString)
+					return nil, false, errors.New(errString)
 				}
-				logCtx.Debugf("Assigned IPs from new block: %s", newIPs)
-				ips = append(ips, newIPs...)
-				rem = num - len(ips)
-				break
+				s.datastoreRetryCount ++
 			}
+			s.datastoreRetryCount ++
 		}
+		return nil, false, errors.New("Max retries hit - excessive concurrent IPAM requests")
+	}
+
+	return nil, false, errors.New("failed to find or claim a block")
+}
+
+func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version int, host string, maxNumBlocks int, rsvdAttr *HostReservedAttr) ([]net.IPNet, error) {
+
+	// First, get the existing host-affine blocks.
+	logCtx := log.WithFields(log.Fields{"host": host})
+	if handleID != nil {
+		logCtx = logCtx.WithField("handle", *handleID)
+	}
+	logCtx.Info("Looking up existing affinities for host")
+	pools, affBlocks, err := c.prepareAffinityBlocksForHost(ctx, requestedPools, version, host, rsvdAttr)
+	if err != nil {
+		return nil, err
+	}
+
+	logCtx.Debugf("Found %d affine IPv%d blocks for host: %v", len(affBlocks), version, affBlocks)
+	ips := []net.IPNet{}
+
+	// Record how many blocks we own so we can check against the limit later.
+	numBlocksOwned := len(affBlocks)
+
+	config, err := c.GetIPAMConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &blockAssignState{
+		client:                c,
+		version:               version,
+		host:                  host,
+		pools:                 pools,
+		remainingAffineBlocks: affBlocks,
+		hostReservedAttr:      rsvdAttr,
+		allowNewClaim:         true,
+	}
+
+assignLoop:
```

Comment:
Can we structure this to remove the need for these loop label markers? In general we've tried to avoid these in Calico since there's usually a better way to represent the same logic. WDYT?